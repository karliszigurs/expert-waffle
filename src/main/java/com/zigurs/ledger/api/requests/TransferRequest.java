package com.zigurs.ledger.api.requests;

import lombok.NonNull;

import java.util.UUID;

public record TransferRequest(@NonNull UUID sourceAccountId,
                              @NonNull UUID destinationAccountId,
                              @NonNull String currency,
                              @NonNull String amount) {
}
