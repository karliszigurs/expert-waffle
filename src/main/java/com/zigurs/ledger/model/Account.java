package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigInteger;
import java.util.UUID;

@Getter
@Entity
@Table(name = "accounts")
@AllArgsConstructor
@NoArgsConstructor
public class Account {

    @Id
    @UuidGenerator
    private UUID id;

    @ManyToOne
    private Client client;

    private String currency;

    @Setter
    private BigInteger balance;
}
