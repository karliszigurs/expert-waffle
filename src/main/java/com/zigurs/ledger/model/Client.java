package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.UuidGenerator;

import java.util.UUID;

@Entity
@Table(name = "clients")
@AllArgsConstructor
@NoArgsConstructor
public class Client {

    @Id
    @UuidGenerator
    @Getter
    private UUID id;
}