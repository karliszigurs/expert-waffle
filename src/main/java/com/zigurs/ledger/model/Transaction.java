package com.zigurs.ledger.model;

import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigInteger;
import java.time.Instant;
import java.util.UUID;

@Getter
@Entity
@Table(name = "transactions")
public class Transaction {

    @Id
    @UuidGenerator
    private UUID id;
    private TransactionStatus status;
    private Instant timestamp;
    @ManyToOne
    private Account sourceAccount;
    private BigInteger sourceAmount;
    private BigInteger sourceBalance;
    @ManyToOne
    private Account destinationAccount;
    private BigInteger destinationAmount;
    private BigInteger destinationBalance;
    private String description;

    public Transaction() {

    }

    public Transaction(UUID id, TransactionStatus status, Instant timestamp, Account sourceAccount, BigInteger sourceAmount, BigInteger sourceBalance, Account destinationAccount, BigInteger destinationAmount, BigInteger destinationBalance, String description) {
        this.id = id;
        this.status = status;
        this.timestamp = timestamp;
        this.sourceAccount = sourceAccount;
        this.sourceAmount = sourceAmount;
        this.sourceBalance = sourceBalance;
        this.destinationAccount = destinationAccount;
        this.destinationAmount = destinationAmount;
        this.destinationBalance = destinationBalance;
        this.description = description;
    }

}
