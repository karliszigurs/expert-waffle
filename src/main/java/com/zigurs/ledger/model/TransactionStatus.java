package com.zigurs.ledger.model;

public enum TransactionStatus {
    COMPLETED,
    FAILED
}
