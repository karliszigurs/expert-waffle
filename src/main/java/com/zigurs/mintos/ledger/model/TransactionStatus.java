package com.zigurs.mintos.ledger.model;

public enum TransactionStatus {
    COMPLETED,
    FAILED
}
