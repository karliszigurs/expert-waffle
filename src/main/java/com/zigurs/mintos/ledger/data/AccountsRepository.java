package com.zigurs.mintos.ledger.data;

import com.zigurs.mintos.ledger.model.Account;

import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface AccountsRepository extends Repository<Account, UUID> {

    List<Account> findByClientId(UUID client_id);

    Optional<Account> findById(UUID uuid);

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("select a from Account a where a.id = :id")
    Optional<Account> findByIdForUpdate(UUID id);

    Account save(Account account);
}
