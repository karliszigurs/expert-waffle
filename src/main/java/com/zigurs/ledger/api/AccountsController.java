package com.zigurs.ledger.api;

import com.zigurs.ledger.api.exceptions.NotFoundException;
import com.zigurs.ledger.api.requests.AccountsRequest;
import com.zigurs.ledger.api.responses.AccountView;
import com.zigurs.ledger.data.AccountsRepository;
import com.zigurs.ledger.data.ClientsRepository;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

/**
 * Accounts controller following Spring Boot patterns.
 */
@RestController
public class AccountsController {

    private final ClientsRepository clientsRepository;
    private final AccountsRepository accountsRepository;

    public AccountsController(
            ClientsRepository clientsRepository,
            AccountsRepository accountsRepository
    ) {
        this.clientsRepository = clientsRepository;
        this.accountsRepository = accountsRepository;
    }

    @PostMapping("accounts")
    public ResponseEntity<List<AccountView>> getAccounts(@RequestBody AccountsRequest request) {
        // Validate client id to differentiate between invalid client vs client, but without accounts.
        clientsRepository.findById(request.client_id()).orElseThrow(
                () -> new NotFoundException("client not found")
        );

        // We know the client id is valid here, we are guaranteed to get at least an empty list
        return ResponseEntity.ok(
                accountsRepository
                        .findByClientId(request.client_id())
                        .stream()
                        .map(AccountView::fromModel)
                        .toList()
        );
    }
}
