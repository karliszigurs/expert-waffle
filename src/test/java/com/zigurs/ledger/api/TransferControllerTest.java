package com.zigurs.ledger.api;

import com.zigurs.ledger.api.requests.TransferRequest;
import com.zigurs.ledger.data.AccountsRepository;
import com.zigurs.ledger.model.Account;
import com.zigurs.ledger.model.Client;
import com.zigurs.ledger.model.Transaction;
import com.zigurs.ledger.model.TransactionStatus;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import testutils.Snapshot;

import java.math.BigInteger;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

import static io.github.jsonSnapshot.SnapshotMatcher.*;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static testutils.Utils.asJsonString;

@WebMvcTest(TransferController.class)
class TransferControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private AccountsRepository accountsRepository;

    @MockBean
    private CurrencyConverterService currencyConverterService;

    @MockBean
    private TransferService transferService;

    @BeforeAll
    public static void beforeAll() {
        start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    void transferHappyPath() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        Instant timestamp = Instant.ofEpochSecond(1715933225L);

        when(transferService.transfer(
                any(),
                eq(sourceAccount.getId()),
                eq(BigInteger.valueOf(5000)),
                eq(destinationAccount.getId()),
                eq(BigInteger.valueOf(5000)),
                anyString()
        )).thenReturn(
                new Transaction(
                        UUID.fromString("424dfead-4625-4dd4-a025-642542eebfdf"),
                        TransactionStatus.COMPLETED,
                        timestamp,
                        sourceAccount,
                        BigInteger.valueOf(100),
                        BigInteger.valueOf(200),
                        destinationAccount,
                        BigInteger.valueOf(300),
                        BigInteger.valueOf(400),
                        "tx one"
                )
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is2xxSuccessful())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferWithConversionHappyPath() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        when(currencyConverterService.convert("EUR", "USD", BigInteger.valueOf(5000))).thenReturn(
                new CurrencyConverterService.CurrencyConversionResult("EUR", "USD", BigInteger.valueOf(7500))
        );

        Instant timestamp = Instant.ofEpochSecond(1715933225L);

        when(transferService.transfer(
                any(),
                eq(sourceAccount.getId()),
                eq(BigInteger.valueOf(7500)),
                eq(destinationAccount.getId()),
                eq(BigInteger.valueOf(5000)),
                anyString()
        )).thenReturn(
                new Transaction(
                        UUID.fromString("424dfead-4625-4dd4-a025-642542eebfdf"),
                        TransactionStatus.COMPLETED,
                        timestamp,
                        sourceAccount,
                        BigInteger.valueOf(100),
                        BigInteger.valueOf(200),
                        destinationAccount,
                        BigInteger.valueOf(300),
                        BigInteger.valueOf(400),
                        "tx one"
                )
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is2xxSuccessful())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferMissingAccount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.empty()
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferMissingDestinationAccount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.empty()
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferSameAccount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferInvalidCurrency() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "USD",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferInvalidAmount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "0"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferMalformedAmount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "I'm a teapot!"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is4xxClientError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferWithConversionInvalidAmount() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        when(currencyConverterService.convert("EUR", "USD", BigInteger.valueOf(5000))).thenReturn(
                new CurrencyConverterService.CurrencyConversionResult("EUR", "USD", BigInteger.valueOf(0))
        );

        Instant timestamp = Instant.ofEpochSecond(1715933225L);

        when(transferService.transfer(
                any(),
                eq(sourceAccount.getId()),
                eq(BigInteger.valueOf(75)),
                eq(destinationAccount.getId()),
                eq(BigInteger.valueOf(50)),
                anyString()
        )).thenReturn(
                new Transaction(
                        UUID.fromString("424dfead-4625-4dd4-a025-642542eebfdf"),
                        TransactionStatus.COMPLETED,
                        timestamp,
                        sourceAccount,
                        BigInteger.valueOf(100),
                        BigInteger.valueOf(200),
                        destinationAccount,
                        BigInteger.valueOf(300),
                        BigInteger.valueOf(400),
                        "tx one"
                )
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is5xxServerError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }

    @Test
    void transferThrowsException() throws Exception {
        Client client = new Client(UUID.randomUUID());

        Account sourceAccount = new Account(
                UUID.fromString("306b74d2-adea-4592-9f8a-981e80a7e041"),
                client,
                "USD",
                BigInteger.valueOf(1L)
        );

        Account destinationAccount = new Account(
                UUID.fromString("1ea6d532-faca-4240-aacf-0356b1e0e950"),
                client,
                "EUR",
                BigInteger.valueOf(1L)
        );

        when(accountsRepository.findById(sourceAccount.getId())).thenReturn(
                Optional.of(sourceAccount)
        );

        when(accountsRepository.findById(destinationAccount.getId())).thenReturn(
                Optional.of(destinationAccount)
        );

        when(currencyConverterService.convert("EUR", "USD", BigInteger.valueOf(50))).thenReturn(
                new CurrencyConverterService.CurrencyConversionResult("EUR", "USD", BigInteger.valueOf(75))
        );

        when(transferService.transfer(
                any(),
                eq(sourceAccount.getId()),
                eq(BigInteger.valueOf(75)),
                eq(destinationAccount.getId()),
                eq(BigInteger.valueOf(50)),
                anyString()
        )).thenThrow(
                new TransferService.TransferException()
        );

        TransferRequest req = new TransferRequest(
                sourceAccount.getId(),
                destinationAccount.getId(),
                "EUR",
                "50"
        );

        MvcResult result = mockMvc.perform(
                        post("/transfer")
                                .content(asJsonString(req))
                                .contentType(MediaType.APPLICATION_JSON)
                                .accept(MediaType.APPLICATION_JSON)
                )
                .andExpect(status().is5xxServerError())
                .andReturn();

        expect(result.getResponse().getContentAsString()).toMatchSnapshot();
    }
}