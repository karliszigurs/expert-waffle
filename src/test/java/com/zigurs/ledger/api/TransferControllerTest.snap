com.zigurs.ledger.api.TransferControllerTest.transferHappyPath=[
  "{\"transaction_id\":\"424dfead-4625-4dd4-a025-642542eebfdf\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"DEBIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"destination_account_id\":\"1ea6d532-faca-4240-aacf-0356b1e0e950\",\"amount\":100,\"account_balance\":200,\"description\":\"tx one\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferInvalidAmount=[
  "{\"error\":\"400 amount must be positive\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferInvalidCurrency=[
  "{\"error\":\"400 currency does not match destination account\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferMalformedAmount=[
  "{\"error\":\"400 amount 'I'm a teapot!' is invalid\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferMissingAccount=[
  "{\"error\":\"404 source_account_id invalid\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferMissingDestinationAccount=[
  "{\"error\":\"404 destination_account_id invalid\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferSameAccount=[
  "{\"error\":\"400 source and destination accounts cannot be the same\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferThrowsException=[
  "{\"error\":\"internal error\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferWithConversionHappyPath=[
  "{\"transaction_id\":\"424dfead-4625-4dd4-a025-642542eebfdf\",\"timestamp\":\"2024-05-17T08:07:05Z\",\"type\":\"DEBIT\",\"status\":\"COMPLETED\",\"source_account_id\":\"306b74d2-adea-4592-9f8a-981e80a7e041\",\"destination_account_id\":\"1ea6d532-faca-4240-aacf-0356b1e0e950\",\"amount\":100,\"account_balance\":200,\"description\":\"tx one\"}"
]


com.zigurs.ledger.api.TransferControllerTest.transferWithConversionInvalidAmount=[
  "{\"error\":\"currency conversion unavailable\"}"
]